### PR TITLE
[#10155] Add sort function to numerical scale summary

### PR DESCRIPTION
**Pull request by Derek-Hardy**
Part of #10155

sort-numerical-scale

**Outline of Solution**
- add the sortable header and icons template
- add sortable data structure to support various sorting functions
- add sort function on click
- add enum to the sort-properties

### CODE REVIEW

File: src/web/types/sort-properties.ts [hunk 1]
```diff
@@ -121,6 +121,26 @@ export enum SortBy {
      * Name of the recipient of the feedback response.
      */
     RECIPIENT_NAME,
+
+    /**
+     * Average score of the numerical scale question.
+     */
+    AVERAGE,
```
Comment (xpdavid) [inline src/web/types/sort-properties.ts:128]: You might want to prefix the properties

File: src/web/app/components/question-types/question-statistics/num-scale-question-statistics.component.ts [hunk 2]
```diff
@@ -1,11 +1,21 @@
...
+import { Component, Input, OnChanges, OnInit } from '@angular/core';
 import {
   FeedbackNumericalScaleQuestionDetails,
   FeedbackNumericalScaleResponseDetails,
 } from '../../../../types/api-output';
 import { DEFAULT_NUMSCALE_QUESTION_DETAILS } from '../../../../types/default-question-structs';
+import { SortBy, SortOrder } from '../../../../types/sort-properties';
 import { QuestionStatistics } from './question-statistics';
 
+interface NumericalScaleStatsRowModel {
```
Comment (xpdavid) [inline src/web/app/components/question-types/question-statistics/num-scale-question-statistics.component.ts:10]: 👍

File: src/web/app/components/question-types/question-statistics/num-scale-question-statistics.component.ts [hunk 3]
```diff
@@ -18,8 +28,21 @@ export class NumScaleQuestionStatisticsComponent
     extends QuestionStatistics<FeedbackNumericalScaleQuestionDetails, FeedbackNumericalScaleResponseDetails>
     implements OnInit, OnChanges {
 
+  // enum
+  SortBy: typeof SortBy = SortBy;
+  SortOrder: typeof SortOrder = SortOrder;
+
+  @Input()
```
Comment (xpdavid) [inline src/web/app/components/question-types/question-statistics/num-scale-question-statistics.component.ts:35]: Why you need it to be `@Input`?

File: src/web/app/components/question-types/question-statistics/num-scale-question-statistics.component.ts [hunk 4]
```diff
@@ -32,6 +55,49 @@ export class NumScaleQuestionStatisticsComponent
...
 
+  sortNumericalScaleStatsRowModel(by: SortBy): void {
+    this.teamToRecipientToScoresSortBy = by;
+    this.teamToRecipientToScoresSortOrder =
+      (this.teamToRecipientToScoresSortOrder === SortOrder.DESC) ? SortOrder.ASC : SortOrder.DESC;
+
+    this.numericalScaleStatsRowModel.sort(this.sortStatsRowBy(by, this.teamToRecipientToScoresSortOrder));
+  }
+
+  sortStatsRowBy(by: SortBy, order: SortOrder):
```
Comment (xpdavid) [inline src/web/app/components/question-types/question-statistics/num-scale-question-statistics.component.ts:66]: Consider using `TableComparatorService`

File: src/web/app/components/question-types/question-statistics/num-scale-question-statistics.component.html [hunk 5]
```diff
@@ -11,23 +11,53 @@
...
-          <td>Recipient</td>
-          <td><span ngbTooltip="Average of the visible responses">Average</span></td>
-          <td><span ngbTooltip="Maximum of the visible responses">Max</span></td>
-          <td><span ngbTooltip="Minimum of the visible responses">Min</span></td>
-          <td><span ngbTooltip="Average of the visible responses excluding recipient's own response to himself/herself">Average excluding self response</span></td>
+          <th class="sortable-header" (click)="sortNumericalScaleStatsRowModel(SortBy.TEAM_NAME)">
+              Team
+              <span *ngIf="teamToRecipientToScoresSortBy !== SortBy.TEAM_NAME"><i class="fas fa-sort"></i></span>
+              <span *ngIf="teamToRecipientToScoresSortBy === SortBy.TEAM_NAME && teamToRecipientToScoresSortOrder === SortOrder.DESC"><i class="fas fa-sort-down"></i></span>
+              <span *ngIf="teamToRecipientToScoresSortBy === SortBy.TEAM_NAME && teamToRecipientToScoresSortOrder === SortOrder.ASC"><i class="fas fa-sort-up"></i></span>
```
Comment (ccyccyccy) [inline src/web/app/components/question-types/question-statistics/num-scale-question-statistics.component.html:18]: Do you think that we can have some sort of component like `SortableTable` that gets rid of the need for all these boilerplate code?

Comment (xpdavid) [reply src/web/app/components/question-types/question-statistics/num-scale-question-statistics.component.html:18]: Good idea. It worth to try and it is tricky to design it generically.

Comment (ccyccyccy) [reply src/web/app/components/question-types/question-statistics/num-scale-question-statistics.component.html:18]: Will try to do this for my MCQ table implementation.

File: src/web/app/components/question-types/question-statistics/num-scale-question-statistics.component.ts [hunk 6]
```diff
@@ -18,9 +29,21 @@ export class NumScaleQuestionStatisticsComponent
     extends QuestionStatistics<FeedbackNumericalScaleQuestionDetails, FeedbackNumericalScaleResponseDetails>
     implements OnInit, OnChanges {
 
+  // enum
+  SortBy: typeof SortBy = SortBy;
+  SortOrder: typeof SortOrder = SortOrder;
+
+  teamToRecipientToScoresSortBy: SortBy = SortBy.NONE;
+
+  @Input()
```
Comment (xpdavid) [inline src/web/app/components/question-types/question-statistics/num-scale-question-statistics.component.ts:38]: The `@Input` here is for?

File: src/web/services/table-comparator.service.ts [hunk 7]
```diff
@@ -11,6 +11,27 @@ export class TableComparatorService {
 
   constructor() { }
 
+  /**
+   * Compares two numbers naturally depending on order given.
+   */
+  compareAsNumber(order: SortOrder, numA: number, numB: number): number {
```
Comment (xpdavid) [inline src/web/services/table-comparator.service.ts:17]: My indentation is actually suggest you to use `compare` in the service. The comparison logic here is essentially the same as `compareNaturally`.

Comment (Derek-Hardy) [reply src/web/services/table-comparator.service.ts]: Initially thought comparing number as string might give unintended behaviour. Did not realise it is handled by option in `.localeCompare()`. 

Updated.

**Pull request merged**